feat: support parse and stringify with comments

lheberlie opened the pull request:
support option to save comments when stringifying

https://github.com/npm/ini/issues/32
https://github.com/npm/ini/issues/83

![image](https://github.com/npm/ini/assets/1824760/d24d7944-1b3d-49f7-8a19-f93d08cd273c)

lheberlie committed:
feat: add support to save comments

parse(data, { preserveComments: true } )
parse(data, { commentsKey: ‘internalComments’, preserveComments: true } )
stringify(data, { preserveComments: true })
stringify(data, { commentsKey: ‘internalComments’, preserveComments: true } )

opt in to save comments when parsing and stringify
save comments to internal object key upon parsing,
add the option to save when stringifying data
store comments internally upon parse and stringify to avoid conflicts
when parsing and stringify multiple documents
use platform eol
include tests in default /test/foo.js

lheberlie committed:
chore: add test for preserving comments

Test option for reading (parse) and saving (stringify) comments in
configuration file.

Test option for reading (parse) and saving (stringify) comments in
configuration file using a custom commentsKey.

lheberlie committed:
docs: add option to preserve comments

Option for reading (parse) and saving (stringify) comments in
configuration file.

lheberlie committed:
fix: simplify commentsRegEx

'{1,}' can be simplified to '+'

lheberlie committed:
fix: code scan on commentsRegEx

Code scanning / CodeQL
Polynomial regular expression used on uncontrolled data
https://codeql.github.com/codeql-query-help/javascript/js-polynomial-redos/

wraithgar commented on lib/ini.js:
```diff
@@ -104,16 +123,30 @@ function splitSections (str, separator) {
+  const commentsKey = opt.commentsKey || 'comments'
+  const commentsDictionary = {}
+  let lineCommentArray = []
+  const commentsRegEx = /^[#;]+[-/\\|~`'":,.;!<>?{}[\]@#$%^&*+=()\w ]+$/
```
This appears to be susceptible to a polynomial ReDOS https://codeql.github.com/codeql-query-help/javascript/js-polynomial-redos/

lheberlie committed:
fix: code scan on commentsRegEx

Code scanning / CodeQL
Polynomial regular expression used on uncontrolled data
https://codeql.github.com/codeql-query-help/javascript/js-polynomial-redos/

wraithgar commented on lib/ini.js:
```diff
@@ -104,16 +123,30 @@ function splitSections (str, separator) {
+  const commentsKey = opt.commentsKey || 'comments'
+  const commentsDictionary = {}
+  let lineCommentArray = []
+  const commentsRegEx = /^[#;]+[^\n]+$/
```
Since we're not doing any grouping or matching, do we really need the `+` after matching `#` or `;`?  They both still would match `[^\n]` anyways.

Also, since the line is already the result of `str.split(/[\r\n]+/g)` I don't understand how it could contain `\n` in the first place.

lheberlie replied:
I guess I was over thinking the solution. I will push an update. Thanks for the feedback.

lheberlie committed:
fix: simplify commentsRegEx

If line starts with comment character ; or #
match line string and push line into line comment array

wraithgar commented on lib/ini.js:
```diff
@@ -104,16 +123,29 @@ function splitSections (str, separator) {
+    if (line && line.match(commentsRegEx)) {
+      lineCommentArray.push(line)
+    }
     if (!line || line.match(/^\s*[;#]/) || line.match(/^\s*$/)) {
```
This line has the regex inline, so the `commentsRegex` can probably be an inline one also.  That's one less line of thought a person has to do when reading the code.

lheberlie replied:
Do you need me to clean all these commits up and submit a new cleaner PR?

wraithgar replied:
No, we squash when we merge so as long as things are heading in the right direction you can just keep pushing changes to your branch.

I know this new PR was based off of suggestions from @lukekarrys so I'll defer to them if this is going in the right direction generally.

wraithgar commented on lib/ini.js:
```diff
@@ -104,16 +123,30 @@
   const duplicates = {}
 
   for (const line of lines) {
+    if (line && line.match(commentsRegEx)) {
```
I wonder if this needs to be regex at all. This is `line.startsWith('#') || line.startsWith(';')`  which avoids all of the polynomial redos problems altogether

wraithgar commented on lib/ini.js:
```diff
@@ -12,6 +12,9 @@ const encode = (obj, opt = {}) => {
   opt.platform = opt.platform || (typeof process !== 'undefined' && process.platform)
   opt.bracketedArray = opt.bracketedArray !== false
+  const commentsKey = opt.commentsKey || 'comments'
+  const commentsKeyRegEx = new RegExp(commentsKey, 'i')
```
Is this really needed?  Building a regex from user input is not the best security practice.  Is it good enough to just say "it's comments" and let folks use that.  Barring that is there a way to define it as a string to match on without regex?

lheberlie committed:
fix: simplify and inline commentsRegEx

lheberlie committed:
fix: remove option to specify commentsKey

Comments will be stored using the ‘comments’ key
Removes security vulnerability

lheberlie committed:
fix: remove test with commentsKey support

wraithgar commented:
Looking good so far, I'll defer to @lukekarrys for the final review since they started it in https://github.com/npm/ini/pull/246.  This probably will not be looked at till next week at the earliest, just a heads up.

harshita2405 commented:
Hi Team and @lukekarrys, any update on this. I also have a similar use case as the above. Do we know when can we expect this to go in ?

lheberlie committed:
fix: remove preserveComments from decode

ini.js
foo.js

lheberlie committed:
fix: preserveComments doc

lheberlie committed:
feat: store comments data as Symbol

parse/decode - store comments as a symbol on object
stringify/encode - restore comments using preserveComments flag

lheberlie commented:
@lukekarrys - sorry it took so long for this update. I hope this will fix the previous issue when using the internal comment key.

![image](https://github.com/user-attachments/assets/03cc0af2-fffa-488f-a857-e1df567787c8)

underwoodsolartechx commented on test/foo.js:
```diff
@@ -94,3 +94,9 @@ test('encode within browser context', function (t) {
   t.matchSnapshot(e)
   t.end()
 })
+test('stringify with comments', function (t) {
```
.decode

underwoodsolartechx reviewed (commented):
sorry

lheberlie commented:
Hi @lukekarrys @wraithgar - just reaching out to see when you might have another chance to review?

wraithgar commented on lib/ini.js:
```diff
@@ -104,16 +123,26 @@ function splitSections (str, separator) {
 const decode = (str, opt = {}) => {
+  // The `typeof` check is required because accessing the `process` directly fails on browsers.
+  /* istanbul ignore next */
+  opt.platform = opt.platform || (typeof process !== 'undefined' && process.platform)
```
I can't think of a compelling reason not to use [`os.EOL`](https://nodejs.org/api/os.html#oseol) instead of adding a new option here

wraithgar replied:
Upon further review this doesn't currently require *any* node modules, so `require('os')` would be a burden to existing browser users.

I also see this is copied from `encode`.  This is ok then.

BlueMystical commented:
Bump!  i also need Comments to be preserved.
Is the project abandoned?

jdossanto158-dot commented on README.md:
```diff
@@ -87,7 +87,10 @@ Attempts to turn the given INI string into a nested data object.
+const object = parse(`<INI Text>`)
+
+// Preserve the comments when parsing
+const objectWithComments = parse(`<INI Text>`, { preserveComments: true }) 
```
cool
/

jdossanto158-dot reviewed (approved):
awesome job

fengmk2 commented:
Is there anything that needs further improvement? I happen to need this feature and hope this pull request can make some progress.

azmy60 reviewed (changes requested):
This doesn't seem to address inline comments. This is the code I used to test it:

```js
const { parse, stringify } = require('./lib/ini');
const config = parse(';database server config\n[server]\nport=8080 ; the database port');
const configStr = stringify(config, { preserveComments: true });
console.log(configStr)
```

The output:

```ini
;database server config
[server]
port=8080
```

lheberlie commented:
I have since moved on from using `ini`, but feel free to take https://github.com/lheberlie/ini/tree/feat-support-parse-stringify-comments and submit another PR.
